src/main/java/org/apache/sling/commons/metrics/internal/JmxExporterFactory.java

```diff
@@ -78,14 +85,46 @@ public class JmxExporterFactory {
     MetricsService metrics;
     
     MBeanServer server;
-    
+
+    NotificationListener listener = new NotificationListener() {
+        public void handleNotification(Notification notification, Object handback) {
+            if (MBeanServerNotification.REGISTRATION_NOTIFICATION.equals(notification.getType())) {
+                ObjectName objectname = null;
+                try {
+                    if(notification instanceof MBeanServerNotification) {
+                        MBeanServerNotification mbeanNotification = (MBeanServerNotification) notification;
+                        objectname = mbeanNotification.getMBeanName();
+                        LOG.debug("JMX Notification : match {} with pattern = {}", objectname, Arrays.asList(patterns));
+                        for (String pattern : patterns) {
+                            if (objectname.toString().matches(pattern)) {
+                                LOG.debug("JMX Notification : register metrics for MBean: {}", objectname);
+                                registerMBeanProperties(objectname);
+                                break;
+                            }
+                        }
+                    } else {
+                        LOG.debug("JMX Notification : Cannot register metrics for objectname = {}", objectname);
+                    }
+                } catch (InstanceNotFoundException | ReflectionException | IntrospectionException e) {
+                    LOG.error("JMX Notification : Cannot register metrics for objectname = {}", objectname, e);
+                }
+            }
+        }
+    };
+
     @Activate
     @Modified
     public void activate(Config config) {
         server = ManagementFactory.getPlatformMBeanServer();
-        registerMetrics(config.objectnames());
+        patterns = config.objectnames();
+        try {
+            server.addNotificationListener(MBeanServerDelegate.DELEGATE_NAME, listener, null, null);
+        } catch (InstanceNotFoundException e) {
+            LOG.error("Cannot add notification listener to MBeanServerDelegate", e);
+        }
+        registerMetrics(patterns);
```

Comment:
I think a comment here would be useful as well, because you (try to) register every MBean here, while at the same time the same MBean can be added by the listener as well. Some clarification (at least that it's safe and does not result in double registrations or something like that) would be useful.

Reply:
Added javadoc comments to explain the behaviour